File: mergeTwoLists.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2){
	struct ListNode* dummy = (struct ListNode*)malloc(sizeof(struct ListNode));
	struct ListNode* cur = dummy;
	while(l1 != NULL && l2 != NULL)
	{
		if(l1->val < l2->val){
			cur->next = l1;
			l1 = l1->next;
		}else if(l1->val >= l2->val){
			cur->next = l2;
			l2 = l2->next;
		}
		cur = cur->next;
	}
	if(l1 != NULL){
		cur->next = l1;
	}else{
		cur->next = l2;
	}

	return dummy->next;
}

int main(int argc, char *argv[]) {
	
	struct ListNode node1 = {1};
	struct ListNode node2 = {2};
	struct ListNode node3 = {4};
	node1.next = &node2;
	node2.next = &node3;
	node3.next = NULL;
	
	struct ListNode node4 = {1};
	struct ListNode node5 = {3};
	struct ListNode node6 = {4};
	node4.next = &node5;
	node5.next = &node6;
	node6.next = NULL;
	
	struct ListNode *list1 = &node1;
	struct ListNode *list2 = &node4;
	struct ListNode *resultList = mergeTwoLists(list1, list2);
	
	while (resultList) {
		printf("%d", resultList->val);
		resultList = resultList->next;
	}
}

File: mergeTwoLists.js
function ListNode(val) {
	this.val = val;
	this.next = null;
}

var mergeTwoLists = function(l1, l2) {
	let dummy = new ListNode();
	let cur = dummy;
	while (l1 && l2) {
		if (l1.val < l2.val) {
			cur.next = l1;
			l1 = l1.next;
		}
		else {
			cur.next = l2;
			l2 = l2.next;
		}
		cur = cur.next;
	}
	
	if (l1) {
		cur.next = l1;
	}
	if (l2) {
		cur.next = l2;
	}
	return dummy.next;
};

function main () {
	
	let node1 = new ListNode(1);
	let node2 = new ListNode(2);
	let node3 = new ListNode(4);
	node1.next = node2; node2.next = node3;

	let node4 = new ListNode(1);
	let node5 = new ListNode(3);
	let node6 = new ListNode(4);
	node4.next = node5; node5.next = node6;

	let resultNode = mergeTwoLists(node1, node4);

	while (resultNode) {
		console.log(resultNode.val);
		resultNode = resultNode.next;
	}
}

main();

